fs-storage/src/folder_storage.rs

```diff
@@ -0,0 +1,336 @@
+            ArklibError::Storage(self.label.clone(), "Key not found".to_owned())
+        })?;
+        // self.ram_timestamps.remove(id);
+        // OR
+        self.ram_timestamps
+            .insert(id.clone(), SystemTime::now());
+        Ok(())
+    }
+
+    /// Compare the timestamp of the storage file
+    /// with the timestamp of the in-memory storage and the last written
+    /// to time to determine if either of the two requires syncing.
+    fn sync_status(&self) -> Result<SyncStatus> {
+        unimplemented!()
+    }
+
+    /// Sync the in-memory storage with the storage on disk
+    fn sync(&mut self) -> Result<()> {
+        match self.sync_status()? {
+            SyncStatus::InSync => Ok(()),
+            SyncStatus::MappingStale => self.read_fs().map(|_| ()),
+            SyncStatus::StorageStale => self.write_fs().map(|_| ()),
+            SyncStatus::Diverge => {
+                let data = self.load_fs_data()?;
+                self.merge_from(&data)?;
+                self.write_fs()?;
+                Ok(())
+            }
+        }
+    }
+
+    /// Read the data from file
+    fn read_fs(&mut self) -> Result<&BTreeMap<K, V>> {
+        let data = self.load_fs_data()?;
+        self.data = data;
+        self.disk_timestamps.clear();
+        for key in self.data.entries.keys() {
+            let file_path = self.path.join(format!("{}.bin", key));
+            if let Ok(metadata) = fs::metadata(&file_path) {
+                if let Ok(modified) = metadata.modified() {
+                    self.disk_timestamps.insert(key.clone(), modified);
+                }
+            }
+        }
+        Ok(&self.data.entries)
+    }
+
+    /// Get a value from the internal mapping
+    fn get(&self, id: &K) -> Option<&V> {
+        self.data.entries.get(id)
+    }
+
+    /// Write the data to file
+    ///
+    /// Update the modified timestamp in file metadata to avoid OS timing issues
+    /// https://github.com/ARK-Builders/ark-rust/pull/63#issuecomment-2163882227
+    fn write_fs(&mut self) -> Result<()> {
+        fs::create_dir_all(&self.path)?;
+
+        for (key, value) in &self.data.entries {
+            let file_path = self.path.join(format!("{}.bin", key));
+            let encoded: Vec<u8> = bincode::serialize(value).map_err(|e| {
+                ArklibError::Storage(
+                    self.label.clone(),
+                    format!("Failed to serialize value: {}", e),
+                )
+            })?;
+
+            let mut file = File::create(&file_path)?;
+            file.write_all(&encoded)?;
+            file.flush()?;
+
+            let new_timestamp = SystemTime::now();
+            file.set_modified(new_timestamp)?;
+            file.sync_all()?;
+
+            self.disk_timestamps
+                .insert(key.clone(), new_timestamp);
+        }
+
+        // Remove files for keys that no longer exist
+        for entry in fs::read_dir(&self.path)? {
```

Comment:
Extract this "clean up" logic into a separate function. It might be useful in other places too.